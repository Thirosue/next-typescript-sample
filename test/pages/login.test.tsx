import React from 'react'
import { act } from 'react-dom/test-utils'
import { render, fireEvent } from '../testUtils'
import Login from '../../pages/login'
import { setCookie, parseCookies, destroyCookie } from 'nookies'
import axios from 'axios'

jest.mock('next/router', () => ({
  useRouter() {
    return {
      push: jest.fn(),
    }
  },
}))

class Page {
  container: HTMLElement
  constructor(container: HTMLElement) {
    this.container = container
  }

  Eメール入力エリア(): Element {
    return this.container.querySelector('#email')
  }

  Eメールエラーメッセージエリア(): Element {
    return this.container.querySelector('.email-error-message-area')
  }

  パスワード入力エリア(): Element {
    return this.container.querySelector('#password')
  }

  パスワードエラーメッセージエリア(): Element {
    return this.container.querySelector('.password-error-message-area')
  }

  オートログインチェックボックス(): Element {
    return this.container.querySelector('#rememberMe')
  }

  確認モーダル(): Element {
    return this.container.querySelector('.modal-dialog')
  }

  確認モーダルタイトル(): Element {
    return this.container.querySelector('.modal-title')
  }

  確認モーダルメッセージ(): Element {
    return this.container.querySelector('.modal-message')
  }

  確認モーダルキャンセルボタン(): Element {
    return this.container.querySelector('.modal-cancel')
  }

  確認モーダルサブミットボタン(): Element {
    return this.container.querySelector('.modal-submit')
  }

  ログインボタン(): Element {
    return this.container.querySelector('.primary-button')
  }

  Email入力(value: string): void {
    fireEvent.change(this.Eメール入力エリア(), { target: { value } })
  }

  パスワード入力(value: string): void {
    fireEvent.change(this.パスワード入力エリア(), { target: { value } })
  }

  オートログインチェック(): void {
    fireEvent.click(this.オートログインチェックボックス())
  }

  確認モーダルキャンセルボタンクリック(): void {
    fireEvent.click(this.確認モーダルキャンセルボタン())
  }

  確認モーダルサブミットボタンクリック(): void {
    fireEvent.click(this.確認モーダルサブミットボタン())
  }

  ログインボタンクリック(): void {
    fireEvent.click(this.ログインボタン())
  }
}

describe('Login page', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Login />, {})
    expect(asFragment()).toMatchSnapshot()
  })

  describe('入力バリデーション', () => {
    it('Eメール未入力で、ログインボタンをクリックしたとき、Eメールが必須エラーとなること', async () => {
      const { container } = render(<Login />, {})
      const page = new Page(container)
      await act(async () => {
        page.ログインボタンクリック()
      })
      expect(page.Eメールエラーメッセージエリア().textContent).toEqual(
        '入力してください'
      )
    })
    it('Eメールをフォーマット誤りで入力のうえ、ログインボタンをクリックしたとき、Eメールがフォーマットエラーとなること', async () => {
      const { container } = render(<Login />, {})
      const page = new Page(container)
      await act(async () => {
        page.Email入力('test@test')
        page.ログインボタンクリック()
      })
      expect(page.Eメールエラーメッセージエリア().textContent).toEqual(
        'メールアドレスを入力してください'
      )
    })
    it('パスワード未入力で、ログインボタンをクリックしたとき、パスワードが必須エラーとなること', async () => {
      const { container } = render(<Login />, {})
      const page = new Page(container)
      await act(async () => {
        page.ログインボタンクリック()
      })
      expect(page.パスワードエラーメッセージエリア().textContent).toEqual(
        '入力してください'
      )
    })
    it('パスワードを桁数不足で入力のうえ、ログインボタンをクリックしたとき、パスワードがフォーマットエラーとなること', async () => {
      const { container } = render(<Login />, {})
      const page = new Page(container)
      await act(async () => {
        page.パスワード入力('Pp1?')
        page.ログインボタンクリック()
      })
      expect(page.パスワードエラーメッセージエリア().textContent).toEqual(
        'アルファベット（大文字小文字混在）と数字と特殊記号を組み合わせて8文字以上で入力してください'
      )
    })
    it('パスワードを組み合わせ不足（記号不足）で入力のうえ、ログインボタンをクリックしたとき、パスワードがフォーマットエラーとなること', async () => {
      const { container } = render(<Login />, {})
      const page = new Page(container)
      await act(async () => {
        page.パスワード入力('Ppppppppppp1')
        page.ログインボタンクリック()
      })
      expect(page.パスワードエラーメッセージエリア().textContent).toEqual(
        'アルファベット（大文字小文字混在）と数字と特殊記号を組み合わせて8文字以上で入力してください'
      )
    })
    it('パスワードを組み合わせ不足（数字不足）で入力のうえ、ログインボタンをクリックしたとき、パスワードがフォーマットエラーとなること', async () => {
      const { container } = render(<Login />, {})
      const page = new Page(container)
      await act(async () => {
        page.パスワード入力('Ppppppppppp?')
        page.ログインボタンクリック()
      })
      expect(page.パスワードエラーメッセージエリア().textContent).toEqual(
        'アルファベット（大文字小文字混在）と数字と特殊記号を組み合わせて8文字以上で入力してください'
      )
    })
    it('パスワードを組み合わせ不足（アルファベット小文字不足）で入力のうえ、ログインボタンをクリックしたとき、パスワードがフォーマットエラーとなること', async () => {
      const { container } = render(<Login />, {})
      const page = new Page(container)
      await act(async () => {
        page.パスワード入力('PXXXXXXXXXXXX1?')
        page.ログインボタンクリック()
      })
      expect(page.パスワードエラーメッセージエリア().textContent).toEqual(
        'アルファベット（大文字小文字混在）と数字と特殊記号を組み合わせて8文字以上で入力してください'
      )
    })
    it('パスワードを組み合わせ不足（アルファベット大文字不足）で入力のうえ、ログインボタンをクリックしたとき、パスワードがフォーマットエラーとなること', async () => {
      const { container } = render(<Login />, {})
      const page = new Page(container)
      await act(async () => {
        page.パスワード入力('ppppppppppppppp1?')
        page.ログインボタンクリック()
      })
      expect(page.パスワードエラーメッセージエリア().textContent).toEqual(
        'アルファベット（大文字小文字混在）と数字と特殊記号を組み合わせて8文字以上で入力してください'
      )
    })
  })

  describe('オートログイン設定', () => {
    it('オートログイン未設定状態でオートログインチェックボックスをチェックしたとき、有効にする旨の確認ダイアログが表示されること', async () => {
      destroyCookie(null, 'rememberMe')
      const { container } = render(<Login />, {})
      const page = new Page(container)
      await act(async () => {
        page.オートログインチェック()
      })
      expect(page.確認モーダル()).toBeTruthy()
      expect(page.確認モーダルメッセージ().textContent).toEqual(
        '自動ログインを有効にしますか？'
      )
    })
    it('オートログイン設定状態でオートログインチェックボックスをチェックしたとき、無効にする旨の確認ダイアログが表示されること', async () => {
      setCookie(null, 'rememberMe', 'true')
      const { container } = render(<Login />, {})
      const page = new Page(container)
      await act(async () => {
        page.オートログインチェック()
      })
      expect(page.確認モーダル()).toBeTruthy()
      expect(page.確認モーダルメッセージ().textContent).toEqual(
        '自動ログインを無効にしますか？'
      )
    })

    it('自動ログイン有効の確認ダイアログが表示され、OKボタンを押したとき、自動ログイン設定がcookieに反映されること', async () => {
      destroyCookie(null, 'rememberMe')
      const { container } = render(<Login />, {})
      const page = new Page(container)
      await act(async () => {
        page.オートログインチェック()
      })
      expect(parseCookies(null).rememberMe).not.toBeTruthy()
      await act(async () => {
        page.確認モーダルサブミットボタンクリック()
      })
      expect(parseCookies(null).rememberMe).toEqual('true')
    })

    it('自動ログイン無効の確認ダイアログが表示され、OKボタンを押したとき、自動ログイン設定がcookieに反映されること', async () => {
      setCookie(null, 'rememberMe', 'true')
      const { container } = render(<Login />, {})
      const page = new Page(container)
      await act(async () => {
        page.オートログインチェック()
      })
      expect(parseCookies(null).rememberMe).toEqual('true')
      await act(async () => {
        page.確認モーダルサブミットボタンクリック()
      })
      expect(parseCookies(null).rememberMe).not.toBeTruthy()
    })
  })

  describe('ログイン確認', () => {
    it('Emailとパスワードに正しいフォーマットを入力の上、ログインボタンを押したとき、認証APIにID/PasswordをPutすること', async () => {
      const { container } = render(<Login />, {})
      const page = new Page(container)
      axios.put = jest.fn().mockImplementation(() =>
        Promise.resolve({
          data: {},
        })
      )
      await act(async () => {
        page.Email入力('test@test.com')
        page.パスワード入力('Password1?')
        page.ログインボタンクリック()
      })
      expect(axios.put).toBeCalledTimes(1)
      expect(axios.put).toBeCalledWith('/api/auth', {
        id: 'test@test.com',
        password: 'Password1?',
      })
    })

    it('ログイン時に、認証APIが認証エラー（401）を返すとき、認証エラーのアラートダイアログが表示されること', async () => {
      const { container } = render(<Login />, {})
      const page = new Page(container)
      axios.put = jest.fn().mockImplementation(() =>
        Promise.reject({
          response: {
            status: 401,
          },
        })
      )
      await act(async () => {
        page.Email入力('test@test.com')
        page.パスワード入力('Password1?')
        page.ログインボタンクリック()
      })
      expect(page.確認モーダル()).toBeTruthy()
      expect(page.確認モーダルタイトル().textContent).toEqual('認証エラー')
      expect(page.確認モーダルメッセージ().textContent).toEqual(
        'Emailもしくはパスワードが誤っています'
      )
    })

    it('ログイン時に、認証APIがシステムエラー（500）を返すとき、システムエラーのアラートダイアログが表示されること', async () => {
      const { container } = render(<Login />, {})
      const page = new Page(container)
      axios.put = jest.fn().mockImplementation(() =>
        Promise.reject({
          response: {
            status: 500,
          },
        })
      )
      await act(async () => {
        page.Email入力('test@test.com')
        page.パスワード入力('Password1?')
        page.ログインボタンクリック()
      })
      expect(page.確認モーダル()).toBeTruthy()
      expect(page.確認モーダルタイトル().textContent).toEqual('システムエラー')
      expect(page.確認モーダルメッセージ().textContent).toEqual(
        'エラーが発生しました。しばらくしてからもう一度お試しください。'
      )
    })
  })
})
