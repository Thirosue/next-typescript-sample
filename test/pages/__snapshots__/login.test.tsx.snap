// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center h-screen bg-gray-200 px-6"
  >
    <div
      class="p-6 max-w-sm w-full bg-white shadow-md rounded-md"
    >
      <div
        class="flex justify-center items-center"
      >
        <svg
          class="h-10 w-10"
          fill="none"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M364.61 390.213C304.625 450.196 207.37 450.196 147.386 390.213C117.394 360.22 102.398 320.911 102.398 281.6C102.398 242.291 117.394 202.981 147.386 172.989C147.386 230.4 153.6 281.6 230.4 307.2C230.4 256 256 102.4 294.4 76.7999C320 128 334.618 142.997 364.608 172.989C394.601 202.981 409.597 242.291 409.597 281.6C409.597 320.911 394.601 360.22 364.61 390.213Z"
            fill="#4C51BF"
            stroke="#4C51BF"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M201.694 387.105C231.686 417.098 280.312 417.098 310.305 387.105C325.301 372.109 332.8 352.456 332.8 332.8C332.8 313.144 325.301 293.491 310.305 278.495C295.309 263.498 288 256 275.2 230.4C256 243.2 243.201 320 243.201 345.6C201.694 345.6 179.2 332.8 179.2 332.8C179.2 352.456 186.698 372.109 201.694 387.105Z"
            fill="white"
          />
        </svg>
        <div>
          <span
            class="text-gray-900 text-2xl font-semibold"
          >
            Dashboard
          </span>
        </div>
      </div>
      <form
        class="mt-4"
      >
        <label
          class="block"
        >
          <div>
            <span
              class="prose prose-sm"
            >
              Email
            </span>
          </div>
          <input
            class="mt-1 w-full border-gray-300 block rounded-md focus:border-indigo-600 "
            id="email"
            name="email"
            type="email"
          />
          <p
            class="text-red-500 text-xs mt-1 email-error-message-area"
          />
        </label>
        <label
          class="block mt-3"
        >
          <div>
            <span
              class="prose prose-sm"
            >
              Password
            </span>
          </div>
          <input
            class="mt-1 w-full border-gray-300 block rounded-md focus:border-indigo-600 "
            id="password"
            name="password"
            type="password"
          />
          <p
            class="text-red-500 text-xs mt-1 password-error-message-area"
          />
        </label>
        <div
          class="flex justify-between items-center mt-4"
        >
          <div>
            <label
              class="inline-flex items-center"
            >
              <input
                class="form-checkbox text-indigo-600"
                id="rememberMe"
                name="rememberMe"
                type="checkbox"
              />
              <div>
                <span
                  class="prose prose-sm mx-2"
                >
                  Remember me
                </span>
              </div>
            </label>
          </div>
          <div>
            <a
              class="block text-sm fontme text-indigo-700 hover:underline"
              href="#"
            >
              Forgot your password?
            </a>
          </div>
        </div>
        <div
          class="mt-6"
        >
          <button
            class="inline-flex justify-center rounded-md border focus:outline-none focus:ring-2 focus:ring-offset-2 text-white bg-indigo-600 hover:bg-indigo-700 focus:ring-indigo-500 border-transparent primary-button py-2 px-4 text-sm w-full "
          >
            Sign in
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
